File: Java/Algorithms/InsertionSort.java
/*Nikhil*/

//Insertion Sort that sort the array within
//the range of [index "fromIndex" to index "inIndex"]
//if you want to run it on whole array
//Simply pass argument OR Replace as follow
//[fromIndex,toIndex] -> [0, ArraySize-1]
//TIME COMPLEXITY - (best,average,worst) -> (n, n^2 , n^2)
//where n: lenght of part of array to be sorted
class InsertionSort 
{
    static void InsertionSort(int[] arr, int fromIndex, int toIndex) 
    {
        for (int i = fromIndex + 1; i <= toIndex; i++) 
        {
            int currValue = arr[i];
            int j = i - 1;

            while (j >= fromIndex && arr[j] > currValue) 
            {
                arr[j + 1] = arr[j];
                j--;
            }
            arr[j + 1] = currValue;
        }
        return;
    }
}

File: Sorting/InsertionSort.cpp
/*Nikhil*/
#include <iostream>
using namespace std;

void insertionSort(int *arr, int size)
{
    for (int i = 1; i <size; i++)
    {
        int currValue = arr[i];
        int j = i - 1;

        while (j >= 0 && arr[j] > currValue)
        {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = currValue;
    }
}


int main()
{
    int arr[] = {2, 9, -1, 5, 10, 1, 0, -7};
    insertionSort(arr, 8);

    //print sorted array
    for (int i = 0; i < 8; i++)
    {
        cout << arr[i] << " ";
    }
    cout << endl;
    return 0;
}
